src/hotspot/cpu/aarch64/aarch64.ad

```diff
@@ -15716,6 +15780,40 @@ instruct minI_rReg(iRegINoSp dst, iRegI src1, iRegI src2)
   %}
 
 %}
+
+instruct minI_imm0_rReg(iRegINoSp dst, immI0 zero, iRegI src2)
+%{
+  match(Set dst (MinI zero src2));
+  ins_cost(INSN_COST * 2);
+  expand %{
+    rFlagsReg cr;
+    compI_reg_imm0(cr, src2);
+    cmovI_reg_imm0_lt(dst, src2, cr);
+  %}
+%}
+
+instruct minI_imm1_rReg(iRegINoSp dst, immI_1 one, iRegI src2)
+%{
+  match(Set dst (MinI one src2));
+  ins_cost(INSN_COST * 2);
+  expand %{
+    rFlagsReg cr;
+    compI_reg_imm0(cr, src2);
+    cmovI_reg_imm1_le(dst, src2, cr);
+  %}
+%}
+
+instruct minI_immM1_rReg(iRegINoSp dst, immI_M1 m1, iRegI src2)
+%{
+  match(Set dst (MinI m1 src2));
+  ins_cost(INSN_COST * 2);
+  expand %{
+    rFlagsReg cr;
+    compI_reg_imm0(cr, src2);
+    cmovI_reg_immM1_lt(dst, src2, cr);
+  %}
+%}
+
```

Comment:
Please put all this repetitive stuff into aarch64_ad.m4 and we'll review that.

Reply:
As min/max ideal transformation https://git.openjdk.org/jdk/pull/9703 looks into child nodes with add and nested min/max, it is tricky to add guarantee of constant being in a fixed input (initial PR variant broke the test ). I rolled the Ideal change back but it leads to doubling of matching rules. So thanks for the advice, that makes even more sense to move things to .m4. I've done that.

CMOV_INSN macro declares instruct-s for generic min/max case with csel.
CMOV_DRAW_INSN macro declares instruct-s that "draw" one of (-1, 0, 1) constants using zr.
MINMAX_DRAW_INSN macro declares instruct-s with matching rules for (-1, 0, 1) immediate operand and a generic register in different order.
There are also helper string macros for upper and lower case to reduce the number of word repetitions.

Existing generic min and max rules were renamed for consistency.